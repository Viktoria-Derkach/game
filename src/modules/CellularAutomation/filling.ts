import { Rules, Cell, Field } from './types';
import { hashRule, demonsRule, venusRule } from './transitions';

export const colors: string[] = [
  '#113',
  '#014',
  '#b16',
  '#54f',
  '#74f',
  '#fff',
  '#add',
  '#4fc',
  '#185',
  '#8c4',
  '#ef6',
  '#4bd',
  '#27f',
  '#33f',
  '#059',
  '#567',
];

export const getColor = (state: Cell) => colors[state];

export const stateColors: number[][] = colors.map((color) => [
  parseInt(color.charAt(1) + color.charAt(1), 16),
  parseInt(color.charAt(2) + color.charAt(2), 16),
  parseInt(color.charAt(3) + color.charAt(3), 16),
  255,
]);

export const getStateColors = (state: Cell) => stateColors[state];

export function randomFill(
  height: number,
  width: number,
  states = colors.length
): Field {
  const result: Field = [];
  for (let y = 0; y < height; y++) {
    const row: Cell[] = [];
    for (let x = 0; x < width; x++) {
      const cell: Cell = Math.floor(Math.random() * states) as Cell;
      row[x] = cell;
    }
    result[y] = row;
  }
  return result;
}

export function transitionFill(
  rule: Rules,
  field: Field,
  height: number,
  width: number,
  states = colors.length
): Field {
  return field.map((row, y, arr) =>
    row.map((cell, x) => {
      const neighbors = {
        top: arr[(y + 1) % height][x],
        right: arr[y][(x + 1) % width],
        bottom: arr[(y + height - 1) % height][x],
        left: arr[y][(x + width - 1) % width],
        rightTop: arr[(y + 1) % height][(x + 1) % width],
        rightBottom: arr[(y + height - 1) % height][(x + 1) % width],
        leftBottom: arr[(y + height - 1) % height][(x + width - 1) % width],
        leftTop: arr[(y + 1) % height][(x + width - 1) % width],
      };
      // CHANGE CELL STATE VIA SELECTED RULE
      if (rule === 'Hash') {
        return hashRule(cell, neighbors, states);
      }
      if (rule === 'Demons') {
        return demonsRule(cell, neighbors, states);
      }
      return venusRule(cell, neighbors);
    })
  );
}
