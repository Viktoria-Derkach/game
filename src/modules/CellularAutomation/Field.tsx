import React, { FC, useEffect, useRef } from 'react';

import { getColor } from './filling';
import { Cell } from './types';

export interface Props {
  field: Cell[][];
  cellSize?: number;
  width: number;
  height: number;
}

export const Field: FC<Props> = ({
  field,
  width,
  height,
  cellSize: cs = 4,
}) => {
  const canvasRef = useRef<HTMLCanvasElement>(null);

  useEffect(() => {
    const canvas = canvasRef.current;
    if (canvas) {
      const context = canvas.getContext('2d');
      if (context) {
        for (let y = 0; y < height; y++) {
          for (let x = 0; x < width; x++) {
            const state = field[y][x];
            context.fillStyle = getColor(state);
            context.fillRect(x * cs, y * cs, cs, cs);
          }
        }
      }
    }
  }, [field, width, height, cs]);

  return (
    <CanvasSection
      canvasRef={canvasRef}
      width={width * cs}
      height={height * cs}
    />
  );
};

const CanvasSection: FC<{
  canvasRef: React.RefObject<HTMLCanvasElement>;
  width: number;
  height: number;
}> = React.memo(({ canvasRef, width, height }) => {
  return (
    <section>
      <canvas ref={canvasRef} width={width} height={height}></canvas>
    </section>
  );
});

CanvasSection.displayName = 'CanvasSection';
