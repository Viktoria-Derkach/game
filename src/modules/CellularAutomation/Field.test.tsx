import React from 'react';
import { render } from '@testing-library/react';

import { Field } from './Field';
import { randomFill } from './filling';

describe('Cellular Automation Field test cases', () => {
  it('Field fragment check', () => {
    const { asFragment } = render(
      <Field width={300} height={300} field={randomFill(300, 300)} />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
