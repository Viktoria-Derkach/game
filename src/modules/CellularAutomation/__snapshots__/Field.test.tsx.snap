// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cellular Automation Field test cases Field fragment check 1`] = `
<DocumentFragment>
  <section>
    <canvas
      height="1200"
      width="1200"
    />
  </section>
</DocumentFragment>
`;
